lark/visitors.py

```diff
@@ -131,11 +134,14 @@ def _transform_tree(self, tree):
         children = list(self._transform_children(tree.children))
         return self._call_userfunc(tree, children)
 
-    def transform(self, tree: Tree) -> _T:
+    def transform(self, tree: Tree[_Leaf_T]) -> _Return_T:
         "Transform the given tree, and return the final result"
         return self._transform_tree(tree)
 
-    def __mul__(self, other: 'Transformer[_T]') -> 'TransformerChain[_T]':
+    def __mul__(
+            self: 'Transformer[Tree[_Leaf_U], _Leaf_T]',
+            other: 'Union[Transformer[_Return_V, _Leaf_U], TransformerChain[_Return_V, _Leaf_U]]'
```

Comment:
This is true. `Transformer` and `TransformerChain` have the same public interface. [Protocol](https://docs.python.org/3/library/typing.html#typing.Protocol) could be used to define that without requiring a base class. (It would just require [typing-extensions](https://pypi.org/project/typing-extensions/) for less than Python 3.8, but that's not a big issue.) I can work on that next.